src/main/java/org/jitsi/videobridge/cc/vp8/VP8FrameMap.java

```diff
@@ -199,32 +205,91 @@ public VP8Frame get(int pictureId)
         public boolean insert(int pictureId, VP8Frame frame)
         {
             int index = indexTracker.update(pictureId);
-            return super.insertItem(frame, index);
+            boolean ret = super.insertItem(frame, index);
+            if (ret)
+            {
+                numCached++;
+                if (firstIndex == -1 || index < firstIndex)
+                {
+                    firstIndex = index;
+                }
+            }
+            return ret;
+        }
+
+        /**
+         * Called when an item in the cache is replaced/discarded.
+         */
+        protected void discardItem(VP8Frame frame)
+        {
+            numCached--;
         }
 
+        /**
+         * Called when the last index of the cache is advanced.
+         */
+        protected void lastIndexAdvanced(int oldlastIndex, int lastIndex)
+        {
+            if (oldlastIndex == -1 || lastIndex == oldlastIndex + 1)
+            {
+                return;
+            }
+            if (lastIndex - oldlastIndex >= getSize())
+            {
+                flush();
+            }
+            for (int i = oldlastIndex + 1; i < lastIndex; i++)
+            {
+                discard(i);
```

Comment:
I wonder if these actually need to be discarded. Can we get away with just masking them in `get()` (i.e. return `null` if we're asked for an old entry even though one exists).

Reply:
I guess that would work.  Is your goal to avoid the linear-time array clearing?

Reply:
Looks like the only reason we explicitly discard is to accurately know the number of cached items, which is only used in the debug output?

Reply:
Yes, and I suppose to reduce the lifetime of Vp8Frame objects once they're not reachable.  But thinking about it the time efficiency is probably more important than the memory efficiency here.